### PR TITLE
Recent Files Menu Bar

### PR DESCRIPTION
**Pull request by ObsidianNA**
RecentsMenuAlpha

Overview: 
Added a Pop-up menu named 'Open Recent' under the 'File' menu bar item that lists 10 recently opened files.

- It stores the paths of opened games inside a file named 'recent.txt' in the same folder where the xenia.exe is located.

- If the file doesn't exits, the program creates it, then paths are stored when a new file is opened.

- Each time a file is opened, the program compares stored paths with the new one, if it exists then the path is moved to the top of the list to indicate that it is the most recent file opened.

- 'Open Recent' items refresh once xenia is re-started

### CODE REVIEW

**Comment by illusion98** [comment]
Should be in a single commit now 😺

File: src/xenia/app/emulator_window.cc [hunk 1]
```diff
@@ -174,10 +182,71 @@ bool EmulatorWindow::Initialize() {
...
+  if (LINES[9] != "" && LINES[9] != "\n")
+    recent_list->AddChild(
+        MenuItem::Create(MenuItem::Type::kString, GLOBAL_RECENT_PATHS[9],
+                         std::bind(&EmulatorWindow::RecentList9, this)));
+//****************************************************************************
   {
     file_menu->AddChild(
         MenuItem::Create(MenuItem::Type::kString, L"&Open...", L"Ctrl+O",
                          std::bind(&EmulatorWindow::FileOpen, this)));
+    // Add Recent List Tab ti file tab
```
Comment (Margen67) [inline src/xenia/app/emulator_window.cc]: *to

File: src/xenia/app/emulator_window.cc [hunk 2]
```diff
@@ -174,10 +182,71 @@ bool EmulatorWindow::Initialize() {
   // FIXME: This code is really messy.
   auto main_menu = MenuItem::Create(MenuItem::Type::kNormal);
   auto file_menu = MenuItem::Create(MenuItem::Type::kPopup, L"&File");
+  // Recent Sub tap
+  auto recent_list = MenuItem::Create(MenuItem::Type::kPopup, L"&Open Recent");
+  //******Add Recent PATHS to menu***************************************
+  std::string LINES[10];
+  std::ifstream RECENT_FILE("recent.txt");
+  int i = 0;
+  // Conver from UTF-8 to whatever tf
```
Comment (Margen67) [inline src/xenia/app/emulator_window.cc]: *Convert

File: src/xenia/app/emulator_window.cc [hunk 3]
```diff
@@ -174,10 +182,71 @@ bool EmulatorWindow::Initialize() {
...
+  std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
```
Comment (gibbed) [inline src/xenia/app/emulator_window.cc]: We have helper functions for this. `xe::to_string`/`xe::to_wstring`.

File: src/xenia/app/emulator_window.cc [hunk 4]
```diff
@@ -174,10 +182,71 @@ bool EmulatorWindow::Initialize() {
...
+      GLOBAL_RECENT_PATHS[i] = converter.from_bytes(LINES[i]);
+      i++;
+    }
+    RECENT_FILE.close();
+  }
+
+  // Added most recently opened files to menu
+  if (LINES[0] != "" && LINES[0] != "\n")
+    recent_list->AddChild(
+        MenuItem::Create(MenuItem::Type::kString, GLOBAL_RECENT_PATHS[0],
```
Comment (gibbed) [inline src/xenia/app/emulator_window.cc]: This should all be a for loop. And std::bind can take static arguments, ie: `std::bind(&EmulatorWindow::RecentList, this, 0)`.

File: src/xenia/app/emulator_window.cc [hunk 5]
```diff
@@ -174,10 +182,71 @@ bool EmulatorWindow::Initialize() {
   // FIXME: This code is really messy.
   auto main_menu = MenuItem::Create(MenuItem::Type::kNormal);
   auto file_menu = MenuItem::Create(MenuItem::Type::kPopup, L"&File");
+  // Recent Sub tap
+  auto recent_list = MenuItem::Create(MenuItem::Type::kPopup, L"&Open Recent");
+  //******Add Recent PATHS to menu***************************************
+  std::string LINES[10];
```
Comment (gibbed) [inline src/xenia/app/emulator_window.cc]: Please follow naming conventions. This applies to all variables and members introduced by your changes. I won't make this comment for every instance.

File: src/xenia/app/emulator_window.cc [hunk 6]
```diff
@@ -174,10 +182,71 @@ bool EmulatorWindow::Initialize() {
...
+  std::ifstream RECENT_FILE("recent.txt");
```
Comment (gibbed) [inline src/xenia/app/emulator_window.cc]: There's things to be said about how you're creating this file here, but I will leave this out until other problems are resolved.

File: src/xenia/app/emulator_window.cc (hunk 4 above)
Comment (ObsidianNA) [reply src/xenia/app/emulator_window.cc]: is the last parameter (0) in 'std::bind(&EmulatorWindow::RecentList, this, 0)' passed to the function?

Comment (gibbed) [reply src/xenia/app/emulator_window.cc]: Yes.

Comment (gibbed) [reply src/xenia/app/emulator_window.cc]: https://en.cppreference.com/w/cpp/utility/functional/bind

File: src/xenia/app/emulator_window.cc (hunk 5 above)
Comment (ObsidianNA) [reply src/xenia/app/emulator_window.cc]: Variable LINES[10] should be lines[10]? Or recent_list to recent_menu?

File: src/xenia/app/emulator_window.cc (hunk 6 above)
Comment (ObsidianNA) [reply src/xenia/app/emulator_window.cc]: ifstream is used to read the file if it exists. If it doesn't then ofstream creates it when a file is opened. Is there still a better way of doing this, if so how?

File: src/xenia/app/emulator_window.cc (hunk 3 above)
Comment (ObsidianNA) [reply src/xenia/app/emulator_window.cc]: Code is using helper functions now

File: src/xenia/app/emulator_window.cc (hunk 5 above)
Comment (ObsidianNA) [reply src/xenia/app/emulator_window.cc]: I read the naming conversion and made the changes. Could you check the code to let me know that there are no more instances of unconventional naming please?

File: src/xenia/app/emulator_window.cc [hunk 7]
```diff
@@ -174,10 +176,37 @@ bool EmulatorWindow::Initialize() {
   // FIXME: This code is really messy.
   auto main_menu = MenuItem::Create(MenuItem::Type::kNormal);
   auto file_menu = MenuItem::Create(MenuItem::Type::kPopup, L"&File");
+  // Recent Sub tab
+  auto recent_list = MenuItem::Create(MenuItem::Type::kPopup, L"&Open Recent");
```
Comment (x1nixmzeng) [inline src/xenia/app/emulator_window.cc]: Minor point, but here the keyboard alt-combo is assigned to alt+o which duplicates open - considering moving this to alt+r (`Open &Recent`)

**Pull request closed without merging**